### DISCUSSION TITLE
Next js  Image  generate additional ampersand in URL

### DISCUSSION THREAD

**Discussion by miko-code** [vercel/next.js#37340] [Help]
We are using next 12.0.4 and we have a weird bug in our production
Sometimes images URL is generated with addition ampersand and causing URL to be malformed
that causing a HTTP 400 error because W parameter cant be parsed correctly. 

The most weird thing is that its happen randomly for some of the traffic.  

```
https://travelpricedrops.com/_next/image?url=%2Fimages%2Fcompanies%2Fexpedia.png%3Fv1.1&amp;w=256&amp;q=75  
```

**Comment by icyJoseph** [comment]
Wow that's wild... The `next/image` default loader, for a while now, AFAIK, ends with:

```ts
  return `${normalizePathTrailingSlash(config.path)}?url=${encodeURIComponent(
    src
  )}&w=${width}&q=${quality || 75}`
```

Why on earth would the ampersand pop up.... I think that line has changed very little over the past 2 years.

I am confused with this one. Do you have a custom loader?

**Reply by miko-code** [reply]
No we don`t use custom loader. 
The app is hosted on Cloud run behind GCP external LB  maybe it related...

Also we tried to change image URL to external  and got same issue.

**Reply by icyJoseph** [reply]
So you can now identify which image is wrong? Yeah it must be that somewhere the url goes through escaping... very rare

**Reply by miko-code** [reply]
Its usually the same group of images that return HTTP 400.
this is  the  image  code 
```
  <Image width={260} height={40}  alt={process.env.NEXT_PUBLIC_SITE_NAME} src={`${process.env.NEXT_PUBLIC_SITE_URL}/images/logos/new-logo.png?v12.1`}
                                           className="Logo"/>
```

**Reply by icyJoseph** [reply]
Where did you pick on this log, or is it from the DOM (src in the img element rendered):

```
https://travelpricedrops.com/_next/image?url=%2Fimages%2Fcompanies%2Fexpedia.png%3Fv1.1&amp;w=256&amp;q=75  
```

This is very strange. I am trying to reproduce it but cannot hit any reproduction. Are you seeing this in any browser in particular?

**Reply by miko-code** [reply]
Its from the cloud run log events
Most of the errors contains those user agents
```
[FBAN/FB4A;FBAV/368.0.0.24.108;FBBV/371898000;FBDM/{density=1.75,width=720,height=1471};FBLC/es_LA;FBRV/373590276;FBCR/#CoberturaCLARO;FBMF/samsung;FBBD/samsung;FBPN/com.facebook.katana;FBDV/SM-A025M;FBSV/11;FBOP/19;FBCA/armeabi-v7a:armeabi;]
AND
Mozilla/5.0 (iPhone; CPU iPhone OS 15_4_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/19E258 [FBAN/FBIOS;FBDV/iPhone14,2;FBMD/iPhone;FBSN/iOS;FBSV/15.4.1;FBSS/3;FBID/phone;FBLC/en_US;FBOP/5]
```

**Reply by miko-code** [reply]
More information , its related to traffic that arrived from   Facebook app   `fbapp://350685531728/unknown`

**Reply by icyJoseph** [reply]
Yeah those added extra user agents are from FB on app browser. Have you been able to try and reliably reproduce this? I'll give it a go this afternoon. I wonder if there's somehow a utm or other sort of tracking system messing up with the url. Or if this is happening in specific devices/browser versions.

**Comment by vadymshymko** [comment]
@miko-code @icyJoseph   Hi there. Is there any updates? @miko-code How did you solve this problem?

**Comment by mathewtrivett** [comment]
Having the same issue on a site I'm working on. Again, it's only certain versions of the Facebook app browser. It looks like the Facebook browser is somehow re-writing / encoding the image URLs? Any thoughts on a workaround or fix for this behaviour?

**Comment by HelaGone** [comment]
I'm not sure if this is related, but I have an issue using the priority prop in next/image. 
As we know setting to true the priority prop will cause a preload of that image. In the `<head>` of the DOM will be a `<link rel="preload"  imageSrcSet="..."  as="image" />` The issue is that those paths are wrongfully encoding the `&` hence causing that image to return a `"w" parameter (width) is required` error. You can try to access this image to see it `https://www.nmas.com.mx/_next/image/?url=https%3A%2F%2Fstatic-live.nmas.com.mx%2Fnmas-news%2Fstyles%2Fcorte_16_9%2Fcloud-storage%2F2024-08%2Fdiscuso-sheinbaum-amlo-reaccion.jpg%3Fh%3D920929c4%26itok%3DCV4-D6A8&amp;w=384&amp;q=80`
I've disabled my middleware to test if that was the problem, but it didn't. Tying to use a custom loader completely disabled the built-in image optimization which is not what I want. 
Anyone here has faced this issue and how to fix it?

**Comment by vigneshMuthumariyappan** [comment]
Hi Guys, I am also facing this same issue @HelaGone faced, On our site during Google Curls for the image goes 404 because of the  **"w" parameter (width) is required** error 

Note:- I am not facing this issue on the actual frontend, It's working fine.

Has anyone here faced this issue and how to fix it?

**Comment by amanpriya-k** [comment]
Hello, I am also facing this issue still.

Did anyone find a solution on this yet? Seeing lots of 400s.

![SCR-20241112-ovpz](https://github.com/user-attachments/assets/2d5ec409-d474-4686-bd73-25b3b3857a76)


---

<img width="500" alt="SCR-20241112-owel" src="https://github.com/user-attachments/assets/c6afb8df-1903-4065-bbf1-81ce5c477771">

**Comment by iDVB** [comment]
We are also seeing this same issue with:
`"w" parameter (width) is required`

Should we **ever** be seeing `&amp;` (HTML entity vs url escaped or left unencoded) in the src of any html? Isn't this a bug?

After chatting it out with Claude Sonnet....

**-- From Claude Sonnet --**

The `&amp;` should never actually be in the URL itself. This is happening because of HTML entity encoding being applied at the wrong level.

Here's the correct sequence that should happen:

1. First, the URL parameters should be properly encoded:
`/_next/image?url=https%3A%2F%2Fimages.ctfassets.net%2Fs5qqrp96y1p4%2F2G79TAYA09J6nVKwlmxa1C%2F9756578f30e4d6a8fbc854ff66cc35ce%2FV2HT_Lab_Coats.jpg&w=3840&q=75`
Note: The url parameter is URL-encoded, but &w= and &q= are clear delimiters
2. This URL should then be placed in the HTML src attribute WITHOUT HTML entity encoding of the & characters. The browser will handle the URL parsing correctly because it sees:
- The main URL parameter delimiter ?
- Clear parameter delimiters &
- URL-encoded content within the url parameter

The presence of `&amp;` suggests that Next.js's Image component is either:
1. Double-encoding the URL (URL encoding + HTML entity encoding), or
2. Applying HTML entity encoding to the entire src attribute value when it should only be encoding special HTML characters that aren't part of the URL structure

This is definitely a bug in Next.js's Image component implementation. The component should:
- URL encode the url parameter value
- Keep the query string delimiters (? and &) as plain characters
- NOT apply HTML entity encoding to URL-specific characters

**Comment by vigneshMuthumariyappan** [comment]
I am just loop the conversation here [https://github.com/vercel/next.js/issues/68998]

@iDVB  Hope the above conversation is helpful 

You are right they are not fully encoded the url

![image](https://github.com/user-attachments/assets/a17bda6e-82df-4e2c-b72e-c59e9f103644)

 Not yet resolved?

**Comment by iDVB** [comment]
Thanks @vigneshMuthumariyappan , 
Will definitly dive into that thread as well.


**Further Claude Sonnet convo thoughts ---**

When you pass a string value to an HTML attribute in React/JSX, React's default HTML escaping kicks in and converts special characters to HTML entities. This is why we see `&amp;` in the rendered HTML - it's React's built-in XSS protection converting `&` to `&amp;` in HTML attribute values.

**This explains why:**

1. The URL parameters work fine in browsers (they un-escape HTML entities before parsing URLs)
2. You see `&amp;` instead of URL-encoded `%26` (it's HTML escaping, not URL encoding)
3. When you manually URL encode the entire string, you get proper URL encoding (`%26`) instead of HTML entities

**The fix in Next.js would need to:**

1. Either URL encode the entire image URL including parameters before passing it to React's attribute system, or
2. Use a different way to construct the URL that doesn't trigger React's HTML escaping

**Reply by icyJoseph** [reply]
Do you have a link to a page that loads an image with the HTML entity?

**Reply by iDVB** [reply]
@icyJoseph It sounds like this is React's built-in XSS protection at work - it automatically escapes special characters in HTML attribute values to prevent injection attacks. When React sees a & in an attribute value, it converts it to &amp; for safety.

We have confirmed that you can actually test this with a simple React example:
`<img src="http://example.com?foo=1&bar=2" />`

This will render as:
`<img src="http://example.com?foo=1&amp;bar=2" />`

So to see a link to a page that loads an image with the HTML entity, you just need a link to any site that uses react and that has an image with at least a single querystring.

Here is one:
https://www.klick.com/news/klick-health-acquires-ward6-singapore

**dev tools element: (`&w=3840&q=75`)** 
This shows that the browser can understand it both ways.... but its our theory that many bots/crawlers can't.
![CleanShot 2025-03-31 at 15 59 41@2x](https://github.com/user-attachments/assets/890fc3cc-d6d0-4546-a9e6-d8860492c366)


**view source: (`&amp;w=3840&amp;q=75`)**
```
<img alt="Singapore Acquisition team" loading="lazy" decoding="async" data-nimg="fill" style="position:absolute;height:100%;width:100%;left:0;top:0;right:0;bottom:0;color:transparent" sizes="(min-width: 912px) 912px, 100vw" srcSet="I REMOVED THIS FOR BREVITY" src="/_next/image?url=https%3A%2F%2Fimages.ctfassets.net%2Fs5qqrp96y1p4%2F1mbrrcC33K2q9wMTQhTDvj%2F8183b3ad1b374e7c60e1d2f300e48566%2FKlick_Health_Singapore_Cropped_Mar16-25.jpg&amp;w=3840&amp;q=75">
```

**Reply by icyJoseph** [reply]
Alright, I see, so react-dom SSR tooling, uses `escapeTextForBrowsers`, which invokes their custom `escapeHTML` function, https://github.com/facebook/react/blob/a7fa8702ee0fc4c04f2d2ff9404e4f626877c113/packages/react-dom-bindings/src/server/escapeTextForBrowser.js#L53

It is not just like, attributes, take for instance:

```jsx
    <span>&</span>
```

The server would send, `<span>&amp;</span>`.

So when React collects the VDOM for a server tree, and makes a simple HTML string out of it, those 5 html entities, get escaped, and then your HTML, as served by the server, has the escaped entities.

Now the thing is... browsers decode these automatically. It is not the only thing where browsers change the incoming HTML, some invalid nesting, for example nested p tags, are re-interpreted, etc... 

So, it'd look like those bots and such, should also process those [character references](https://developer.mozilla.org/en-US/docs/Glossary/Character_reference)

**Comment by iDVB** [comment]
@icyJoseph Sounds like you have a better handle on the issue then I.
We're not even sure **what** is trying to reference the urls without the "w" parameter, but its ALOT. 
We're getting like 20-30 requests a day to these 500 error resulting next/image urls. (next image function returns 500, feels like it should return a 4xx vs 5xx)

**Reply by icyJoseph** [reply]
Do you have logs on user agents arriving on the request?

**Reply by iDVB** [reply]
user agents I have! 
Seems like mostly iOS, maybe to do with the browser "peeking/preloading" links when hovered?

```
Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1 RDDocuments/8.4.8.940

Mozilla/5.0 (iPhone; CPU iPhone OS 17_0_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1 OPT/4.3.1

Mozilla/5.0 (iPhone; CPU iPhone OS 17_6_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.6 Mobile/15E148 Safari/604.1

Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36 Edg/121.0.0.0

Mozilla/5.0 (iPad; CPU OS 17_0_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/605.1.15

Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:123.0) Gecko/20100101 Firefox/123.0
```

**Reply by icyJoseph** [reply]
🤔 but how could we reproduce this feature? It looks like it has to be a bot/crawler that, gets the site HTML, and straight up parses the string, without fixing the character references.

**Comment by vigneshMuthumariyappan** [comment]
@icyJoseph Simply, Why we shouldn't encode full url instead  of concat &w={..} as mentioned in the below image 🤔? 

Is there cause any issue?

![image](https://github.com/user-attachments/assets/4d2d396d-994e-4324-98d6-ec1eb82b4436)

**Reply by icyJoseph** [reply]
Those are query parameters, the url is encoded so that it can be transported as a query parameter as well.

The issue that has been pointed out, is that, when doing the SSR string, ReactDOM invokes a function that replaces 5 character references, that is when it finds `&`, it swaps it for `&amp;`, and that is what you'd see in the string served by the server. Not only in Next Images, for text inside tags, etc...  

- https://github.com/facebook/react/blob/a7fa8702ee0fc4c04f2d2ff9404e4f626877c113/packages/react-dom-bindings/src/server/escapeTextForBrowser.js#L53
